feat: Annnotation based app prefix detection in webhooks

carsonoid opened the pull request:
Checklist:

* [X] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [X] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Optional. My organization is added to USERS.md.
* [ ] I've signed the CLA and my build is green ([troubleshooting builds](https://argoproj.github.io/argo-cd/developer-guide/ci/)).

wtam2018 commented on util/webhook/webhook.go:
```diff
@@ -160,40 +182,95 @@ func (a *ArgoCDWebhookHandler) HandleEvent(payload interface{}) {
 
+func getAppRefreshPrefix(app *v1alpha1.Application) string {
+	// an explicitPrefix setting always takes precence
+	if explicitPrefix := app.Annotations["argocd.argoproj.io/refresh-prefix"]; explicitPrefix != "" {
```
I would define all annotations in common.go as constants and document each of them in const's comments.

wtam2018 reviewed (commented):
Is there an issue created for this feature?    Please add more description about this feature.

carsonoid commented:
@wtam2018 thanks for the review! We are having a discussion about this feature in #4272.

Right now it is up in the air as to whether this feature will be allowed in. As soon as I get buy-off on adding the feature I will make the changes you suggested :)

carsonoid commented:
@wtam2018 suggestions implemented. I have also added docs to explain the annotations.

alexmt commented:
Hello @carsonoid , thanks a lot for working on this change! The repo server performance is one of the main pain points that affects Argo CD scalability. 

The one concern about this PR is that it only postpones manifest generation by a few minutes. It seems like a semi-solution so we were hesitating to introduce more annotations to enable it. Argo CD already has so many configuration knobs and we are adding more very carefully. 

I would like to suggest an enhancement. In addition to preventing application refresh, the webhook handler could "pre-populate" manifest cache for the new commit using path annotation and previously generated manifests cache. So for example we have a repo with one application in the directory `guestbook`, the `http://argocd.argoproj.io/refresh-on-path-updates-only` annotation is set to true and the repository has one commit `a`. If webhook receive new commit `b` that changes only 'README.md' then refresh can be skipped and additionally handle can copy cached manifests of commit `a` into the commit `b` cache.  This would `cancel` the manifests generation.

What do you think?

alexmt commented:
Probably this change would be really easy to implement. We have API to get and set manifest cache:

https://github.com/argoproj/argo-cd/blob/9f47a116219700a46750707fd4d0e5f835404444/reposerver/cache/cache.go#L71-L77

alexmt commented on common/common.go:
```diff
@@ -123,6 +123,19 @@ const (
+	// it tells the api server to only trigger app refreshes when the webhook contains changes for
+	// files under the given path.
+	// The associated value should be a file path.
+	AnnotationKeyRefreshPrefix = "argocd.argoproj.io/refresh-prefix"
```
Couple suggestions here. 
* Rename annotation to make more explicit that it is related to manifest generation. E.g. `argocd.argoproj.io/manifest-generation-path`.


* Can we treat value in this annotation as a relative path and support `.` value? E.g. 

```yaml
annotations:
  argocd.argoproj.io/refresh-prefix: guestbook
```

In this case, user can add `argocd.argoproj.io/refresh-prefix: .` and we probably don't need `argocd.argoproj.io/refresh-on-path-updates-only` at all.

jessesuen commented:
@carsonoid we'd like to continue your effort on this, but with some few changes (including renaming annotation and implementing some of alexes suggestions). Can we take what you've done so far and put it in an argo-cd branch where we could continue the effort?

carsonoid commented:
@jessesuen yeah that's totally fine by me. Sorry for the slow response. I have been out of town for a few a while and unable to make the suggested changes yet.

I would say that all of the suggestions from @alexmt are good, except maybe the relative path. Unless you plan on letting users do relative paths behind the app path. The main reason for even having the `refresh-prefix` annotation was to specifically handle cases where `kustomize` was dependent on files outside of the `Application` path. `Ex ../../../base`. So the relative would need to support directory traversal in order to still cover that particular use-case.

alexmt commented:
Good point about relative path @carsonoid ! To address it the `argocd.argoproj.io/manifest-generation-path` should support `../my-shared-base` in the path as well as value might be absolute path within the repo `/my-shared-base`.

I've created another PR that includes your commits and addresses comments of this PR. https://github.com/argoproj/argo-cd/pull/4699

carsonoid commented:
That's great! Thanks for picking this up and running with it.

On Wed, Oct 28, 2020 at 2:06 PM Alexander Matyushentsev <
notifications@github.com> wrote:

> Good point about relative path @carsonoid <https://github.com/carsonoid>
> ! To address it the argocd.argoproj.io/manifest-generation-path should
> support ../my-shared-base in the path as well as value might be absolute
> path within the repo /my-shared-base.
>
> I've created another PR that includes your commits and addresses comments
> of this PR. #4699 <https://github.com/argoproj/argo-cd/pull/4699>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/argoproj/argo-cd/pull/4273#issuecomment-718178396>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAWV6SKZSIOVVFEHNOYHLWLSNB2VPANCNFSM4QZVEFAQ>
> .
>

alexmt commented:
https://github.com/argoproj/argo-cd/pull/4699 incorporated changes from this PR
